src/MSRIOGroup.cpp

```diff
@@ -550,6 +576,31 @@ namespace geopm
         }
     }
 
+    bool MSRIOGroup::is_hwp_enabled(void)
+    {
+        uint32_t leaf = 6; //thermal and power management features
+        uint32_t eax, ebx, ecx, edx;
+        const uint32_t hwp_mask = 0x80;
+        bool supported, enabled = false;
+        __get_cpuid(leaf, &eax, &ebx, &ecx, &edx);
+
+        supported = (bool)((eax & hwp_mask) >> 7);
+
+        if (supported) {
+            int domain = signal_domain_type("MSR::PM_ENABLE:ENABLE");
+            int num_domain = m_platform_topo.num_domain(domain);
+            double pkg_enable = 0.0;
+            for (int dom_idx = 0; dom_idx < num_domain; ++dom_idx) {
+                pkg_enable += read_signal("MSR::PM_ENABLE:ENABLE", domain, dom_idx);
+            }
+            if (pkg_enable == 1.0*num_domain) {
+                enabled = true;
```

Comment:
pkg_enable to bool

Reply:
This should be changed to throw an error if all packages aren't enabled to use HWP or Legacy p-states.